wrapper/telemetry.go

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache v2.0 License.
+
+//
+// This file provides telemetry support for wrapper functions
+//
+
+package telemetry
+
+import (
+	"context"
+	"regexp"
+	"runtime"
+	"strings"
+
+	"github.com/Microsoft/go-winio/pkg/etw"
+	"github.com/Microsoft/go-winio/pkg/guid"
+	"github.com/microsoft/moc-sdk-for-go/services/admin/version"
+	"github.com/microsoft/moc/pkg/auth"
+)
+
+// GUID: {37009198-5B9B-4B40-B1A9-6C561FB69D26}
+var providerGUID guid.GUID = guid.FromArray([16]byte{
+	0x37, 0x00,
+	0x91, 0x98,
+	0x5B, 0x9B,
+	0x4B, 0x40,
+	0xB1, 0xA9,
+	0x6C, 0x56,
+	0x1F, 0xB6,
+	0x9D, 0x26,
+})
+
+const ProviderName string = "Microsoft.AKSHCI.MocSdkForGo.Wrapper"
+const WrapperTelemetryVersion string = "v1.0.1"
+
+type WrapperEventFields struct {
+	EventName         string
+	CorrelationVector string
+	FunctionName      string
+	Subroutine        string
+	ErrorString       string
+	MocVersion        string
+	MocAgentVersion   string
+}
+
+// serverName is used when requesting MocVersion info and is NOT to be emitted in telemetry.
+func EmitWrapperTelemetry(eventName string, correlationVector string, errorString string, subroutine string, serverName string) {
+
+	// calling for version information is costly, only perform at login
+	mocVersion, mocAgentVersion := "", ""
+	if eventName == "SecurityLoginCV" {
+		mocVersion, mocAgentVersion = getMocVersion(serverName)
+	}
+	functionName := ""
+	var details *runtime.Func
+	pc, _, _, ok := runtime.Caller(1)
+	if ok {
+		details = runtime.FuncForPC(pc)
+	}
+	if ok && details != nil {
+		functionName = details.Name()
+	}
+
+	fields := WrapperEventFields{
+		eventName,
+		correlationVector,
+		functionName,
+		subroutine,
+		FilterSensitiveData(errorString),
+		mocVersion,
+		mocAgentVersion,
+	}
+	fields.emitWrapperTelemetry()
+}
+
+func (eventTelemetry WrapperEventFields) emitWrapperTelemetry() {
+
+	provider, err := etw.NewProviderWithOptions(ProviderName, etw.WithID(providerGUID))
+	if err != nil {
+		return
+	}
+	defer provider.Close()
+
+	err = provider.WriteEvent(
+		eventTelemetry.EventName,
+		etw.WithEventOpts(),
+		etw.WithFields(
+			etw.StringField("CorrelationVector", eventTelemetry.CorrelationVector),
+			etw.StringField("Function", eventTelemetry.FunctionName),
+			etw.StringField("Subroutine", eventTelemetry.Subroutine),
+			etw.StringField("Error", eventTelemetry.ErrorString),
+			etw.StringField("WrapperTelemetryVersion", WrapperTelemetryVersion),
+			etw.StringField("MocVersion", eventTelemetry.MocVersion),
+			etw.StringField("MocAgentVersion", eventTelemetry.MocAgentVersion),
+		))
+	if err != nil {
+		return
+	}
+}
+
+func getMocVersion(serverName string) (string, string) {
```

Comment:
I dont see where this is used?
Note i think this is internal to the wrapper. You can return err as well

Reply:
Do you mean `getMocVersion`? That is used on `telemetry.go` line 53: `mocVersion, mocAgentVersion = getMocVersion(serverName)`